home/vasco-santos/2021-03-exploration-report-1-on-js-lotus-client.md

```diff
@@ -0,0 +1,102 @@
+# Experiment with lotus client
+
+## Deal Flow
+
+Requisites:
+- A wallet id with FIL
+
+Flow:
+1. import file
+  * `provider.importFile(data: String): Promise<CID>`
+2. fetch file info for deal
+  * `client.clientDealSize(cid: Cid): Promise<DataSize>`
+3. list miners
+  * `client.stateListMiners(tipSetKey: Cid[]): Promise<Array<string>>`
+4. get miner info for each miner
+  * `client.stateMinerInfo(address: string, tipSetKey: Cid[]): Promise<MinerInfo>`
+5. find the price and conditions of each miner
+  * `client.clientQueryAsk(id: string, address: string): Promise<StorageAsk>`
+6. start a deal with the intended miners
+  * `clientStartDeal (startDealParams: StartDealParams): Promise<Cid>`
+
+
+```js
+// const wallet = ...
+const cidString = await provider.importFile('this is random text!')
+
+const cid = {
+  '/': cidString
+}
+
+const dataSize = await client.clientDealSize(cid)
+
+const head = await client.chainHead()
+const miners = await client.stateListMiners(head.Cids)
+
+for (const miner in miners) {
+  let minerInfo, storageAsk
+  try {
+    minerInfo = await client.stateMinerInfo(miner, head.Cids)
+    storageAsk = await client.clientQueryAsk(minerInfo.PeerId, miner)
+  } catch (err) {
+    console.log('error trying to get conditions for miner', miner)
+  }
+  
+  // TODO: add your logic for: should we try this minor?
+
+  const dealCid = await client.clientStartDeal({
+    Data: {
+      PieceCid: cid,
+      PieceSize: dataSize.PieceSize,
+      Root: cid,
+      TransferType: ''
+    },
+    DealStartEpoch: -1, // default
+    EpochPrice: storageAsk.Price,
+    FastRetrieval: true,
+    MinBlocksDuration: undefined,
+    Miner: selectedMiner,
+    ProviderCollateral: undefined,
+    VerifiedDeal: false,
+    Wallet: wallet
+  })
+}
+```
+
+Notes:
+- the expectations here would be that `client.clientDealSize` returned the actual pieceSize to provide, but I am getting: `unexpected deal status while waiting for data request: 11 (StorageDealFailing). Provider message: deal rejected: proposal piece size is invalid: padded piece size must be a power of 2`
+- this is the [recommended flow](https://docs.filecoin.io/store/lotus/store-data), but for the lotus client we should probably leave the file import for the end, once we have a miner ready to receive the file. We need to compute the size first though.
```

Comment:
When making the deal there needs to be a match of CommP between the client and miner, I believe this is part of the deal interaction from early in the process—so that when the miner receives the data they can confirm it is what they were told it is. CommP is basically a complicated hash of the contents and for anything of sufficient size it's very inefficient to do it in JavaScript. So getting the data into Lotus probably needs to happen before starting the deal process to it can handle all of this I think.

Reply:
My reasoning here is what happens to lots of content that might be imported to a lotus node and then the deal is not effective, as well as the time to transfer a large chunk of data just to negotiate a deal. 
However, we should also have a solution for clean up when the deal is successful, which I am not sure we have